pilot/pkg/networking/core/waypoint.go

```diff
@@ -48,39 +47,37 @@ type waypointServices struct {
 func findWaypointResources(node *model.Proxy, push *model.PushContext) ([]model.WorkloadInfo, *waypointServices) {
 	network := node.Metadata.Network.String()
 	workloads := make([]model.WorkloadInfo, 0)
+	serviceInfos := make([]model.ServiceInfo, 0)
 	for _, svct := range node.ServiceTargets {
 		ips := svct.Service.ClusterVIPs.GetAddressesFor(node.GetClusterID())
-		wl := push.WorkloadsForWaypoint(model.WaypointKey{
+		key := model.WaypointKey{
 			Network:   network,
 			Addresses: ips,
-		})
+		}
+		wl := push.WorkloadsForWaypoint(key)
 		workloads = append(workloads, wl...)
+		svcs := push.ServicesForWaypoint(key)
+		serviceInfos = append(serviceInfos, svcs...)
 	}
 
-	return workloads, findWorkloadServices(workloads, push)
-}
-
-func findWorkloadServices(workloads []model.WorkloadInfo, push *model.PushContext) *waypointServices {
-	wps := &waypointServices{}
-	for _, wl := range workloads {
-		for _, ns := range push.ServiceIndex.HostnameAndNamespace {
-			svc := ns[wl.Namespace]
-			if svc == nil {
-				continue
-			}
-			if labels.Instance(svc.Attributes.LabelSelectors).Match(wl.Labels) {
-				if wps.services == nil {
-					wps.services = map[host.Name]*model.Service{}
-				}
-				wps.services[svc.Hostname] = svc
-			}
+	waypointServices := &waypointServices{}
+	for _, s := range serviceInfos {
+		hostName := host.Name(s.Service.Hostname)
+		svc, ok := push.ServiceIndex.HostnameAndNamespace[hostName][s.Namespace]
+		if !ok {
+			continue
+		}
+		if waypointServices.services == nil {
+			waypointServices.services = map[host.Name]*model.Service{}
 		}
+		waypointServices.services[hostName] = svc
```

Comment:
Though not introduced from this pr., wanna to know the case when we can have multi services associated with a waypoint.

Reply:
(sorry for the late response, I missed this question)
I am not fully sure I understand the question so please LMK if I miss your point. This should be services that are handled/guarded by the waypoint so 1 example would be namespace is annotated to use some waypoint and there are multiple services in the namespace.

Reply:
As the param node of this function  is the waypoint itself, i am not sure how can we set its targetServices to the real services it proxies.

You see in L51~L60, only service targets are added  to `serviceInfos`

Reply:
There is an index which maps the VIP (plus network technically) of the gateway to the services (and service-adjacent types) which use the waypoint. We are using this to do the lookup.